lib/view_component/base.rb

```diff
@@ -99,13 +99,12 @@ def render?
 
     def initialize(*); end
 
-    # If trying to render a partial or template inside a component,
-    # pass the render call to the parent view_context.
+    # use original view_context if we're not rendering a component
```

Comment:
```suggestion
    # Re-use original view_context if we're not rendering a component.
    # 
    # This prevents an exception when rendering a partial inside of a component that has also been rendered outside
    # of the component. This is due to the partials compiled template method existing in the parent `view_context`,
    #  and not the component's `view_context`.
```

Thoughts on something like the above, explaining why this is necessary?

Reply:
This change tries not to interfere with the way Rails works, thus naturally should not exhibit unexpected results. In that sense it may not be necessary to explain in long form?

Is there a specific reason to render with the component as context? Implementing it like this passes all tests (and, in my Rails app it doesn't seem to lead to immediate problems):

``` ruby
    # Render using the original `view_context` Rails provides.
    #
    # This makes sure that compiled template methods live on that `view_context`
    # and not on the rendering component.
    def render(options = {}, args = {}, &block)
      view_context.render(options, args, &block)
    end
```

It makes sure we don't interfere with Rails expecting its methods to be defined within its `view_context`. On the other hand, the render monkey patch makes sure components can render as expected.

Not committing immediately as I may not see the reason why the `view_context` needs to change.

Reply:
@BlakeWilliams oh, and thanks for reviewing! Really appreciate view_component and your fast responses!

Reply:
> Is there a specific reason to render with the component as context? Implementing it like this passes all tests (and, in my Rails app it doesn't seem to lead to immediate problems):

That's a good question. I'm not 100% sure, but I could see it being a performance reason.

Without diving into it, I think re-using the current `view_context` via `super` avoids instantiating a new view context instance, which is done when calling `view_context`.

There's a small benchmark script that can be run via `bundle exec rake benchmark`, we could verify that assumption by running that against `master` and this branch.

Reply:
Well, ran it more than once, that's for sure :-).

I always get the message:

````
config.eager_load is set to nil. Please update your config/environments/*.rb files accordingly:

  * development - set it to false
  * test - set it to false (unless you use a tool that preloads your test environment)
  * production - set it to true
````

Always using `view_context` (twice):

````
Warming up --------------------------------------
          component:    16.155k i/100ms
            partial:     1.691k i/100ms
Calculating -------------------------------------
          component:    174.574k (± 5.6%) i/s -      1.745M in  10.029241s
            partial:     16.601k (± 6.8%) i/s -    165.718k in  10.034281s

Comparison:
          component::   174574.1 i/s
            partial::    16601.4 i/s - 10.52x  (± 0.00) slower


			
Warming up --------------------------------------
          component:    16.032k i/100ms
            partial:     1.614k i/100ms
Calculating -------------------------------------
          component:    177.213k (± 6.6%) i/s -      1.764M in  10.004032s
            partial:     16.234k (± 6.7%) i/s -    163.014k in  10.094963s

Comparison:
          component::   177213.0 i/s
            partial::    16234.1 i/s - 10.92x  (± 0.00) slower		
````
			

Master (changing context):

````

Warming up --------------------------------------
          component:    15.924k i/100ms
            partial:     1.712k i/100ms
Calculating -------------------------------------
          component:    174.103k (± 6.3%) i/s -      1.736M in  10.013919s
            partial:     16.490k (± 8.6%) i/s -    164.352k in  10.051274s

Comparison:
          component::   174103.2 i/s
            partial::    16489.8 i/s - 10.56x  (± 0.00) slower
			
			
Warming up --------------------------------------
          component:    14.450k i/100ms
            partial:     1.495k i/100ms
Calculating -------------------------------------
          component:    171.259k (± 9.0%) i/s -      1.705M in  10.051173s
            partial:     14.743k (± 9.9%) i/s -    146.510k in  10.055019s

Comparison:
          component::   171259.2 i/s
            partial::    14743.0 i/s - 11.62x  (± 0.00) slower			
````

Original version of this branch (changing context only for view components):

````

Warming up --------------------------------------
          component:    15.924k i/100ms
            partial:     1.712k i/100ms
Calculating -------------------------------------
          component:    174.103k (± 6.3%) i/s -      1.736M in  10.013919s
            partial:     16.490k (± 8.6%) i/s -    164.352k in  10.051274s

Comparison:
          component::   174103.2 i/s
            partial::    16489.8 i/s - 10.56x  (± 0.00) slower
			
			
Warming up --------------------------------------
          component:    14.450k i/100ms
            partial:     1.495k i/100ms
Calculating -------------------------------------
          component:    171.259k (± 9.0%) i/s -      1.705M in  10.051173s
            partial:     14.743k (± 9.9%) i/s -    146.510k in  10.055019s

Comparison:
          component::   171259.2 i/s
            partial::    14743.0 i/s - 11.62x  (± 0.00) slower			
````

Reply:
Thanks! It looks like we do have a small performance penalty when we don't share the `view_context` between component renders. Given that, I think we'll want to continue avoiding `view_context.render` for components.

Reply:
Oh, right, I'll commit an explanation there. I don't really see the advantage of using a component specific context, it deviates from Rails default creating the origin of this issue and performs somewhat worse. Also note the lower standard deviation in the performance test. It's consistently lower than the one on master.
That said, using another context may fit better for reasons I don't know anything about so I'll commit an explanation making this ready for a merge.